Write tutorial for graphql (#66)
Knotx/knotx-website: src/render/tutorials/graphql-usage/2_0.html.md

---
title: Using GraphQL with Knot.x
author: Voycawojka
keywords: tutorial
date: 2019-10-16
layout: tutorial
knotxVersions:
  - 2.0.0
---
## Overview

> In this tutorial we'll implement a sample GraphQL endpoint that makes use of Knot.x tasks system. It's advised to complete the [Getting Started tutorials](/tutorials/getting-started-with-knotx-stack/2_0/) first. You should also be familiar with fragments. Familiarity with GraphQL is a nice-to-have but it's not necessary.

Author: I believe you meant "not obligatory". I'll change it like this :D

You'll learn how to:
- use java and Vert.x GraphQL implementations
- implement them to work seamlessly with Knot.x
- create and use custom fragments

## GraphQL

> GraphQL is a protocol, a way of serving and retrieving data. It has similar applications to REST but provides a few advantages over it. Unlike REST, GraphQL uses a single endpoint that is serving all data. Clients specify what data they need with a dedicated query language. Server then responds with JSON including only data the client asked for and in the structure he wanted.

Author: I meant that it can be used as an alternative to it. But I see it was confusing so the sentence is gone

At runtime, you assign data fetchers to types defined in graphlQL schema. Such a fetcher needs to return an object that can be automaticaly mapped to the schema type. For example, if the schema defines an operation `book` returning a type `Book` that has a property `title` then we would implement it like that:

1. Define a POJO `Book` object

```java
class Book {
    private String title;

    public Book(String title) {
        this.title = title;
    }
}
```

2. Define a fetcher

```java
class BookFetcher implements DataFetcher<Book> {
    @Override
    Book get(DataFetchingEnvironment env) {
        return new Book("Reindeers World");
    }
}
```

The book is hardcoded for the example's sake. Normaly it would probably call some 3rd party API. But as far as GraphQL is concerned, it doesn't matter as long as the fetcher returns a `Book` object. 

3. Assign the fetcher to operation

```java
//...
RuntimeWiring runtimeWiring = newRuntimeWiring()

>     .type("QueryType", builder -> builder.dataFetcher("book", new BookFetcher()))

Reviewer: Could you add a few words about `"QueryType"`?

        }
    }
}
```

Quite simple.

## Vert.x Web GraphQL module
Note that the above java implemention has nothing to do with HTTP. This is where the Vert.x implementation comes in handy. Vert.x provides an out of the box HTTP handler for GraphQL. We just need to provide it with a configured `GraphQL` object and it's ready to serve requests through HTTP.

## Get the best out of GraphQL and Knot.x
There are two steps to integrate GraphQL and Knot.x. The first is to define a GraphQL handler factory that will return a `GraphQLHandler` (the one from the Vert.x implementation). It will allow us to define and configure GraphQL endpoint in the Knot.x configuration.

We could stop at implementing the factory but we wouldn't be using the full power of Knot.x. Knot.x has some great mechanisms for _fetching_ data from 3rd parties. Therefore, the second step is to create data fetchers that make use of Knot.x integration functionalities.

> It's assumed you have the project set up and ready for development. If you have troubles with that check the [Getting Started tutorials](/tutorials/getting-started-with-knotx-stack/2_0/).

## Step 0 - Google Books API, schema and model

> Let's stick with books. We'll implement a sample service serving books from Google Books API. Google Books is a typical REST API. It returns a lot of information. In this tutorial we'll extract just a title, publisher and list of authors (google api returns authors as a list of strings). The response from the Google API looks like this:

Reviewer: ```suggestion
Let's stick with books. We'll implement a sample service serving books from Google Books API. Google Books is a typical REST API. It returns a lot of information. In this tutorial we'll extract just a title, publisher and a list of authors (google api returns authors as a list of strings). The response from the Google API looks like this:
```

> ```json5

Reviewer: ```suggestion
```json
```

Author: It's on purpose. JSON5 is an extension of JSON. It supports comments that I use in this snippet. Knotx website treats it as a regular JSON, but github for instance supports JSON5.

So yea, it's a json5 snippet

{
    "items": [
        {
            //...
            "volumeInfo": {
                "title": "...",
                "publisher": "...",
                "authors": ["Author1", "Author2"],
                //...
            }
        }
    ]
}
```

Our schema should look like this:

```graphql
schema {
    query: QueryType
}

type QueryType {
    books(match: String): [Book]
    book(id: String): Book
}

type Book {
    title: String!
    publisher: String!
    authors: [String]
}
```

It defines two operations:
- `books`: takes a string parameter (search keyword) and returns a list of books
- `book`: takes a string parameter (google id) and returns a single book

The schema also defines what a book is. Exclamation point means a field is mandatory (can't be null).

Save it as a `*.graphqls` file in your app resources.

> We'll also need to implement a java model for every type our operations can return. It needs to match the schema exacly. In this case it's just one class, Book:

Reviewer: ```suggestion
We'll also need to implement a java model for every type our operations can return. It needs to match the schema exacly. In this case it's just one class, `Book`:
```

```java
class Book implements GraphQLDataObject {
    private String title;
    private String publisher;
    private List<String> authors;

    @Override
    public void fromJson(JsonObject json, DataFetchingEnvironment environment) {
        JsonObject volumeInfo = json.getJsonObject("volumeInfo");

        title = volumeInfo.getString("title");
        publisher = volumeInfo.getString("publisher");
        authors = new LinkedList<>();

        volumeInfo.getJsonArray("authors", new JsonArray()).forEach(object -> authors.add((String) object));
    }
}
```

Note that it's implementing `GraphQLDataObject` interface with a `fromJson` method. It will come in handy later. It's not a standard interface, you have to define it:

```java
public interface GraphQLDataObject extends Serializable {
  void fromJson(JsonObject json, DataFetchingEnvironment environment);
}
```

## Step 1 - The factory

> Let's implement the factory that will setup a `GraphQL` object and produce a Vert.x `GraphQLHandler`.

Reviewer: ```suggestion
Let's implement the [Handler](link to documentation) factory that will setup a `GraphQL` object and produce a Vert.x `GraphQLHandler`.
```

}
```

Rememeber about updating services in META-INF. With a handler factory you can now configure a graphql endpoint. Add this path to `openapi.yml`:

```yml
/api/graphql:
    post:
        operationId: get-books-gql
        responses:
            default:
                description: Books Graphql API
```

Note that it needs to be POST because GraphQL reads queries from the requests' bodies. Now add the operation to `operations.conf`:

```HOCON
{
    operationId = get-books-gql
    handlers = ${config.server.handlers.common.request} [
        {
            name = graphqlHandler
            config = {
                schema = "books.graphqls"
            }
        }
    ]
}
```

Note the `${config.server.handlers.common.request}` part. We import some standard configuration because we'll need it later. Now the only thing left are data fetchers and you'll have a working graphql endpoint.

## Step 2
We would like to fetch data from 3rd parties using existing Knot.x mechanisms, not to reinvent the wheel. Ideally we would have a generic data fetcher that can process Knot.x tasks defined in configuration. 

> Our fetcher will parse configuration to create tasks and run them using `FragmentsEngine`. It will output a `Fragment` that will contain data fetched from google in it's payload. Fragment's payload is just a `JsonObject` where actions can store data.

Reviewer: please add documentation links

#### Execute tasks from configuration

> Let's start with something like this:

Reviewer: please provide some description of this `something` ;)

Author: This `something` is explained in details below the code snippet

```java
public abstract class TaskDataFetcher<T> implements DataFetcher<CompletableFuture<T>> {

    private final Vertx vertx;
    private final JsonObject config;
    private final RoutingContext routingContext;
    private final String taskName;
    private final FragmentsEngine engine;

    TaskDataFetcher(VertX, JsonObject config, RoutingContext routingContext, String taskName) {
        this.vertx = vertx;
        this.config = config;
        this.routingContext = routingContext;
        this.taskName = taskName;
        engine = new FragmentsEngine(vertx);
    }

    @Override
    public CompletableFuture<T> get(DataFetchingEnvironment environment) {
        FragmentEventContextTaskAware eventContextTaskAware = setupTask(vertx, config, routingContext, environment);
        CompletableFuture<T> future = new CompletableFuture<>();

        engine
            .execute(Collections.singletonList(eventContextTaskAware))
            .subscribe(events -> {
                // Here we'll get the outcome fragment's payload, parse it to model and complete the future with it
            });
        
        return future;
    }

>     //other methods we'll implement in a moment

Reviewer: ```suggestion
    //we'll implement other methods in a moment
```

Author: Is it really better? To be on the same page, in my version there is an implied "that":

>other method (that) we'll implement in a moment

Reviewer: I didn't hear that `that` in my head so it sounded quite odd :/

}
```

> There's a couple of things to explain here. Firstly, our fetcher is abstract - it will have one abstract method for transforming payload into objects from our model (`Book`).

Reviewer: ```suggestion
There's a couple of things to explain here. First, our fetcher is abstract - it will have one abstract method for transforming payload into objects from our model (`Book`).
```

Author: I think `firstly` is fine. For reference: https://grammarist.com/usage/firstly-secondly-thirdly/

Reviewer: Technically you're right. It just sounded weird to me. But I don't insist on changing that

Also, it's generic. That's because it doesn't matter for it what type will it return - it may be `Book`, it may be an array `Book[]` or it may be anything else.

`TaskDataFetcher` implements `DataFetcher` (every data fetcher needs to) of type `CompletableFuture<T>`. Java implementation of GraphQL allows us to return futures instead of plain objects. If a data fetcher returns a future, graphQL will asyncronicly wait for it to complete. Only when all futures returned by data fetchers complete, the graphql will construct a response.

> Our abstract fetcher creates an instance of `FragmentsEngine`. The engine can execute task trees. Also, in the `get` method we create `FragmentEventContextTaskAware`. It's basicaly a task tree ready to be executed by the engine. Let's define the `setupTask` method that will take a configuration and parse it into a task tree. We'll also need one constant.

Reviewer: ```suggestion
Our abstract fetcher creates an instance of `FragmentsEngine`. The engine can execute task graphs. Also, in the `get` method we create `FragmentEventContextTaskAware`. It's basicaly a task graph ready to be executed by the engine. Let's define the `setupTask` method that will take a configuration and parse it into a task graph. We'll also need one constant.
```

Author: Trees are specific cases of graphs. The engine operates on task trees. "Graphs" could imply that cycles are possible.

Reviewer: you're right but afaik we use the word `graph` everywhere else (in docs etc.)

    Fragment fragment = new Fragment(FRAGMENT_TYPE, fragmentConfig, "");

    RequestContext requestContext = routingContext.get(RequestContext.KEY);
    ClientRequest clientRequest = requestContext.getRequestEvent().getClientRequest();

    FragmentEvent event = new FragmentEvent(fragment);
    FragmentEventContext eventContext = new FragmentEventContext(event, clientRequest);

    FragmentsHandlerOptions options = new FragmentsHandlerOptions(config);
    ActionProvider proxyProvider = new ActionProvider(options.getAction(), supplyFactories(), vertx.getDelegate());
    TaskBuilder taskBuilder = new TaskBuilder(FRAGMENT_TYPE, options.getTasks(), proxyProvider);

    Task task = taskBuilder
        .build(fragment)
        .orElseThrow(() -> new NullPointerException("No task built"));

    return new FragmentEventContextTaskAware(task, eventContext);
}
```

Let's take a look at it piece by piece:

```java
JsonObject fragmentConfig = new JsonObject();
fragmentConfig.put(FRAGMENT_TYPE, taskName);
fragmentConfig.put("gql", new JsonObject(env.getArguments()));

Fragment fragment = new Fragment(FRAGMENT_TYPE, fragmentConfig, "");
```

> We create a new fragment that will be processed by tasks. It needs to know which task should start processing it, therefore `taskName` is assigned to its type. We also put graphql environment arguments to the configuration under "gql" key. Environment arguments are just values passed to parameters of operation. In our case "match" in `books(match: String)` and "id" in `book(id: String)`. By putting it into fragment configuration, we allow tasks to access it. Note that the "gql" key is arbitrary and you can call it anything you want. 

Reviewer: ```suggestion
We create a new fragment that will be processed by tasks. It needs to know which task should start processing it, therefore `taskName` is assigned to its type. We also put GraphQL environment arguments to the configuration under "gql" key. Environment arguments are just values passed to parameters of an operation. In our case "match" in `books(match: String)` and "id" in `book(id: String)`. By putting it into a fragment configuration, we allow tasks to access it. Note that the "gql" key is arbitrary and you can call it anything you want. 
```

For the third argument in the fragment's contructor we pass an empty string. It's our fragment's body which we're not interested in in this tutorial.

```java
RequestContext requestContext = routingContext.get(RequestContext.KEY);
ClientRequest clientRequest = requestContext.getRequestEvent().getClientRequest();
```

We need to extract `ClientRequest` from `routingContext`. Remember the `${config.server.handlers.common.request}` part in the operation configuration in `operations.conf`? Without it there wouldn't be anything in `routingContext` under `RequestContext.KEY`.

```java
FragmentEvent event = new FragmentEvent(fragment);
FragmentEventContext eventContext = new FragmentEventContext(event, clientRequest);
```

Our final task tree will need two things: `Task` and `FragmentEventContext`. Here we create the latter. It keeps the client request and our fragment in the form of `FragmentEvent`.

```java
FragmentsHandlerOptions options = new FragmentsHandlerOptions(config);
ActionProvider proxyProvider = new ActionProvider(options.getAction(), supplyFactories(), vertx.getDelegate());
TaskBuilder taskBuilder = new TaskBuilder(FRAGMENT_TYPE, options.getTasks(), proxyProvider);
```

> In order to create a task we need to use `TaskBuilder`. It needs to be supplied with a few things from the configuration. We use `FragmentsHandlerOptions` to parse the configuration. We also need a supplier of action factories. We can retrieve it using standard java `ServiceLoader`. We'll take a look at `supplyFactories` method in a moment.

Reviewer: ```suggestion
In order to create a task we need to use a `TaskBuilder`. It needs to be supplied with a few things from the configuration. We use a `FragmentsHandlerOptions` to parse the configuration. We also need a supplier of action factories. We can retrieve it using the standard java `ServiceLoader`. We'll take a look at `supplyFactories` method in a moment.
```

Author: I was wondering about that. For example here:

>We use `FragmentsHandlerOptions` to parse the [...]

I think `FragmentsHandlerOptions` works like a proper noun and therefore don't need an article. (Just like you just say "Poland" or "Bob")

Reviewer: maybe you're right, I think we use it like that everywhere else
It just sounded right to do it here, so maybe for the sake of keeping it concise, let's leave it like that

Author: I've adressed that ✍️

```java
@Override
public CompletableFuture<T> get(DataFetchingEnvironment environment) {
    FragmentEventContextTaskAware eventContextTaskAware = setupTask(vertx, config, routingContext, environment);
    CompletableFuture<T> future = new CompletableFuture<>();

    engine
        .execute(Collections.singletonList(eventContextTaskAware))
        .subscribe(events -> {
            JsonObject payload = events.get(0).getFragment().getPayload();
            JsonObject fetchedData = payload.getJsonObject("fetchedData");
            T model = getDataObjectFromJson(fetchedData, environment);
            future.complete(model);
        });

    return future;
}
```

After the engine is done executing the task tree we:
- retrieve the fragment and its payload
- get data from payload under "fetchedData" key
- transform this data into our model (in our case it would be `Book`)
- complete the future with our model so that GraphQL gets the data

> There are two things to explain here. Firstly, we assume that the data we need is under "fetchedData" key. It is not exacly the case but we'll deal with it later. Secondly, we use `getDataObjectFromJson` method to transform json into our model. This method is abstract because our `TaskDataFetcher` don't know how to do it.

Reviewer: ```suggestion
There are two things to explain here. Firstly, we assume that the data we need is under "fetchedData" key. It is not exactly the case but we'll deal with it later. Secondly, we use `getDataObjectFromJson` method to transform json into our model. This method is abstract because our `TaskDataFetcher` doesn't know how to do it.
```

> There are two things to explain here. Firstly, we assume that the data we need is under "fetchedData" key. It is not exacly the case but we'll deal with it later. Secondly, we use `getDataObjectFromJson` method to transform json into our model. This method is abstract because our `TaskDataFetcher` don't know how to do it.

Reviewer 2: There are few repetitions of sentences like:
> There are things to explain here...

Maybe it would be worth to add some alternatives like:
> So what just happened? We assumed...
> Some explanation to the code above...
> What we did above was...

One more thing to explain here is `toArray` parameter passed to our fetcher. It's a `Function` used to transform `JsonObject` from payload into a `JsonArray`. In our case, google api returns something like:

```json5
{
    "items": [
        {/* book 1 */},
        {/* book 2 */ },
        //...
    ]
}
```

Therefore, we would pass the following function as `toArray`: 

```java
json -> json.getJsonArray("items")
```

#### Using our fetchers
We can now assign our fetchers to operations in runtime wiring builder in `GraphQLHandlerFactory`:

```java
RuntimeWiring runtimeWiring = newRuntimeWiring()
        .type("QueryType", builder -> builder
            .dataFetcher("books", new TaskArrayDataFetcher<>("get-books", Book.class, vertx, config, routingContext, json -> json.getJsonArray("items")))
            .dataFetcher("book", new TaskSingleDataFetcher<>("get-book", Book.class, vertx, config, routingContext))
        )
        .build();
```

> We just assign operation `books` to a `TaskArrayDataFetcher` that uses task `get-books` from configuration and returns an array of `Book`. And we assign operation `book` to a `TaskSingleDataFetcher` that uses task `get-book` and returns a `Book`.

Reviewer: ```suggestion
We just assign operation `books` to a `TaskArrayDataFetcher` that uses task `get-books` from configuration and returns an array of `Book` objects. Also, we assign operation `book` to a `TaskSingleDataFetcher` that uses task `get-book` and returns a `Book`.
```

> Let's set it up in Knot.x configuration. It's best to extract our graphql configuration to a seperate file. The graphql operation configuration in `operations.conf` should now look like this:

Reviewer: ```suggestion
Let's set it up in Knot.x configuration. It's best to extract our GraphQL configuration to a separate file. The GraphQL operation configuration in `operations.conf` should now look like this:
```

```HOCON
{
    operationId = get-books-gql
    handlers = ${config.server.handlers.common.request} [
        {
            name = graphqlHandler
            config = {include required(classpath("routes/handlers/graphqlHandler.conf"))}
        }
    ]
}
```

Now we can define graphql configuration in a new file, `routes/handlers/graphqlHandler.conf`:

```HOCON
schema = "books.graphqls"

tasks {
  get-books {
    action = getBooks
  }

  get-book {
    action = getBook
  }
}

actions {
  getBooks {
    factory = http
    config {
      endpointOptions {
        path = "/books/v1/volumes?q={config.gql.match}"
        domain = www.googleapis.com
        port = 443
        allowedRequestHeaders = ["Content-Type"]
      }
      webClientOptions {
        ssl = true
      }
    }
  }

  getBook {
    factory = http
    config {
      endpointOptions {
        path = "/books/v1/volumes/{config.gql.id}"
        domain = www.googleapis.com
        port = 443
        allowedRequestHeaders = ["Content-Type"]
      }
      webClientOptions {
        ssl = true
      }
    }
  }
 
} 
```

> We simply define two tasks (one per graphql operation, we reference their names in fetcher constructors) each with one HTTP action. Those actions call appropriate endpoints:

Reviewer: ```suggestion
We simply define two tasks (one per each GraphQL operation, we reference their names in fetcher constructors) each with one HTTP action. Those actions call appropriate endpoints:
```

- `/books/v1/volumes?q={config.gql.match}` to get books by a given keyword
- `/books/v1/volumes/{config.gql.id}` to get a single book with a given id

Note the `{config.gql.xxx}` parts. At runtime its swapped for values from the fragment configuration (remember, we put environment arguments in fragment's config under "gql" key). For example, if we request data like this: 

```graphql
{
    book(id: "cool_id") {
        title
    }
}
```

Then the http action will swap `{config.gql.id}` for "cool_id" and call `/books/v1/volumes/cool_id`.

#### Exposing the data under the right payload key

> There is just one last thing we need to do. Every http action saves retrieved data in payload under the action's name and a "\_result" subkey. `getBook` action will save the data under "getBook.\_result" key and `getBooks` action will save the data under "getBooks._result" key. Remember that our fetcher assumes the data is under "fetchedData" key. Therefore, we need an action that will transfer data in payload from one key to the other. We'll call it "expose-payload-data".

Reviewer: ```suggestion
There is just one last thing we need to do. Every http action saves retrieved data in payload under the action's name and a "\_result" subkey. `getBook` action will save the data under "getBook.\_result" key and `getBooks` action will save the data under "getBooks.\_result" key. Remember that our fetcher assumes the data is under "fetchedData" key. Therefore, we need an action that will transfer data in payload from one key to the other. We'll call it "expose-payload-data".
```

Add two new actions to `graphqlHandler.conf`:

```HOCON
exposeInPayload-getBooks {
    factory = expose-payload-data
    config {
        key = getBooks
        exposeAs = fetchedData
    }
}

exposeInPayload-getBook {
    factory = expose-payload-data
    config {
        key = getBook
        exposeAs = fetchedData
    }
}
```

And then setup transitions in our tasks. They should now look like this:

```HOCON
tasks {
  get-books {
    action = getBooks
    onTransitions {
      _success {
        action = exposeInPayload-getBooks
      }
    }
  }

  get-book {
    action = getBook
    onTransitions {
      _success {
        action = exposeInPayload-getBook
      }
    }
  }
}
```

## We're done

That's it. You can now run the application. Easiest way of seeing it in action is using postman, but you can use curl or anything else you want.

Let's construct a sample query:

```graphql
{
    book(id: "q5NoDwAAQBAJ") {
        title
        authors
    }
    books(match: "java") {
        publisher
    }
}
```

> We're asking for the title and authors of the book with id "q5NoDwAAQBAJ" and a list of books about java with only their publishers listed. Request it at http://localhost:8092/api/graphql

Reviewer: ```suggestion
We're asking for the title and authors of the book with id `"q5NoDwAAQBAJ"` and a list of books about java with only their publishers listed. We'll send our request to `http://localhost:8092/api/graphql`
```

If you're using postman you can just choose predefined `GraphQL` body type and paste the query there. If you prefer curl:

```bash
curl -i -H 'Content-Type: application/json' -X POST -d '{"query": "{book(id: \"q5NoDwAAQBAJ\") {title authors} books(match: \"java\") {publisher}}"}' http://localhost:8092/api/graphql
```

You should get the following response:

```json5
{
    "data": {
        "book": {
            "title": "Learning GraphQL",
            "authors": [
                "Eve Porcello",
                "Alex Banks"
            ]
        },
        "books": [
            {
                "publisher": "Helion"
            },
            {
                "publisher": "Helion"
            },
            {
                "publisher": "Helion"
            },
            {
                "publisher": "\"O'Reilly Media, Inc.\""
            },
            {
                "publisher": "Morgan Kaufmann"
            },
            //...
        ]
    }
}
```

## Few notes at the end

> As you can see Knot.x is highly customizable. Data fetchers that we implemented are quite generic. You can use them to easily create more advanced GraphQL APIs without the need for writing more fetchers. For clarity purposes the code in this tutorial don't have advanced error handling and is quite error-prone. It's left as an exercise for you to implement better error handling.

Reviewer: ```suggestion
As you can see Knot.x is highly customizable. Data fetchers that we implemented are quite generic. You can use them to easily create more advanced GraphQL APIs without the need for writing more fetchers. For clarity purposes the code in this tutorial doesn't have advanced error handling and is quite error-prone. It's left as an exercise for you to implement better error handling.
```

> As you can see Knot.x is highly customizable. Data fetchers that we implemented are quite generic. You can use them to easily create more advanced GraphQL APIs without the need for writing more fetchers. For clarity purposes the code in this tutorial don't have advanced error handling and is quite error-prone. It's left as an exercise for you to implement better error handling.

Reviewer 2: ```suggestion
As you can see Knot.x is highly customizable. Data fetchers that we implemented are quite generic. You can use them to easily create more advanced GraphQL APIs without the need for writing more fetchers. For clarity purposes the code in this tutorial doesn't have advanced error handling and is quite error-prone. It's left as an exercise for you to implement better error handling.
```

> Also, you might have noted that our `fromJson` method in `GraphQLDataObject` interface have a `DataFetchingEnvironment` argument that we didn't use. This is because our example is very simple, but the argument is included in code to show the flexibility of our solution. You can imagine a more complex case where we need to parse json to model differently, dependending on query parameters. This is where we could use our `DataFetchingEnvironment` argument.

Reviewer 2: ```suggestion
Also, you might have noted that our `fromJson` method in `GraphQLDataObject` interface has a `DataFetchingEnvironment` argument that we didn't use. This is because our example is very simple, but the argument is included in code to show the flexibility of our solution. You can imagine a more complex case where we need to parse JSON to model differently, depending on query parameters. This is where we could use our `DataFetchingEnvironment` argument.
```

---
title: Using GraphQL with Knot.x
author: Voycawojka
keywords: tutorial
date: 2019-10-16
layout: tutorial
knotxVersions:
  - 2.0.0
---
## Overview
In this tutorial, we'll implement a sample [GraphQL](https://graphql.org/) endpoint that makes use of Knot.x tasks system. It's advised to complete the [Getting Started tutorials](/tutorials/getting-started-with-knotx-stack/2_0/) first. You should also be familiar with fragments. Familiarity with GraphQL is not obligatory (but always welcome).

You'll learn how to:
- use Java and Vert.x GraphQL implementations
- implement them to work seamlessly with Knot.x
- create and use custom fragments

## GraphQL

> GraphQL is a protocol, a way of serving and retrieving data. Unlike REST, GraphQL uses a single endpoint that is serving all data. Clients specify what data they need with a dedicated query language. The server then responds with JSON including only data the client asked for and in the structure he or she wanted.

Author: Right

        employer
    }
}
```

And get a JSON response:

```json
{
    "data": {
        "reindeers": [
            {
                "name": "Rudolph",
                "employer": "Santa Claus"
            }
        ]
    }
}
```

Maybe we could ask the server for different properties of reindeers or for other animals - it doesn't matter. We get what we asked for.

Another advantage of GraphQL is that it's statically typed. The server defines a schema which is a file defining all operations and types the client can request. It's like a model layer.

You can read more at [graphql.org](https://graphql.org/).

## Java implementation library
Because GraphQL is a specification, it has many different implementations in many different languages. The most popular Java one is [GraphQL Java](https://github.com/graphql-java/graphql-java).

The main idea of this library are the **data fetchers**. A data fetcher is an object implementing the `DataFetcher` interface with only one method: 

```java
T get(DataFetchingEnvironment env)
```

Let's ignore the `env` argument for now. As you might suspect, the role of a data fetcher is to provide some data (return an object, to be precise).

> At runtime, you assign data fetchers to types defined in GraphQL schema. Such a fetcher needs to return an object that can be automatically mapped to the schema type. For example, if the schema defines an operation `book` returning a type `Book` that has a property `title`, then we would implement it like that:

Author: I don't remember, but I think someone suggested this comma. I don't want to change it back and forth so I'll just leave it here

We could stop at implementing the factory but we wouldn't be using the full power of Knot.x. Knot.x has some great mechanisms for _fetching_ data from 3rd parties. Therefore, the second step is to create data fetchers that make use of Knot.x integration functionalities.

> It's assumed you have the project set up and ready for development. If you have troubles with that check the [Getting Started tutorials](/tutorials/getting-started-with-knotx-stack/2_0/).

### Step 0 - Google Books API, schema and model
Let's stick with books. We'll implement a sample service serving books from the Google Books API. Google Books is a typical REST API. It returns a lot of information. In this tutorial, we'll extract just a title, publisher and a list of authors (Google API returns authors as a list of strings). The response from the Google API looks like this:

```json5
{
    "items": [
        {
            //...
            "volumeInfo": {
                "title": "...",
                "publisher": "...",
                "authors": ["Author1", "Author2"],
                //...
            }
        }
    ]
}
```

> 

Reviewer 3: Add example API queries:
```
curl https://www.googleapis.com/books/v1/volumes\?q\=Java
curl https://www.googleapis.com/books/v1/volumes/UEdjAgAAQBAJ
```

Author: done

Our schema should look like this:

```graphql
schema {
    query: QueryType
}

type QueryType {
    books(match: String): [Book]
    book(id: String): Book
}

type Book {
    title: String!
    publisher: String!
    authors: [String]
}
```

It defines two operations:
- `books`: takes a string parameter (search keyword) and returns a list of books
- `book`: takes a string parameter (google id) and returns a single book

The schema also defines what a book is. The exclamation point means a field is mandatory (can't be null).

> Save it as a `books.graphqls` file in your app resources.

Reviewer 3: Maybe specify a catalogue here.

    return GraphQL.newGraphQL(graphQLSchema).build();
  }

private static Reader loadResource(String path) {
    return new InputStreamReader(GraphQLHandlerFactory.class.getResourceAsStream("/" + path));
}
```

Remember about updating services in the `META-INF`. With a handler factory, you can now configure a GraphQL endpoint. Add this path to `openapi.yml`:

```yml
/api/graphql:
    post:
        operationId: get-books-gql
        responses:
            default:
                description: Books Graphql API
```

Note that it needs to be POST because GraphQL reads queries from the requests' bodies. Now add the operation to `operations.conf`:

```HOCON
{
    operationId = get-books-gql
    handlers = ${config.server.handlers.common.request} [
        {
            name = graphqlHandler
            config = {
                schema = "books.graphqls"
            }
        }
    ]
}
```

Note the `${config.server.handlers.common.request}` part. We import some standard configuration because we'll need it later. Now the only thing left are data fetchers and you'll have a working GraphQL endpoint.

### Step 2 - Data fetchers and FragmentsEngine

> We would like to fetch data from 3rd parties using existing Knot.x mechanisms, not to reinvent the wheel. Ideally we would have a generic data fetcher that can process Knot.x tasks defined in configuration. 

Reviewer 3: ```suggestion
We would like to fetch data from 3rd parties using existing Knot.x mechanisms, not to reinvent the wheel. Ideally we would have a generic data fetcher that can process Knot.x [Tasks](https://github.com/Knotx/knotx-fragments/tree/2.0.0/handler#how-does-it-work) defined in configuration. 
```

> Our fetcher will parse configuration to create tasks and run them using [`FragmentsEngine`](https://github.com/Knotx/knotx-fragments/tree/master/handler/engine). It will output a `Fragment` that will contain data fetched from google in its payload. Fragment's payload is just a `JsonObject` where actions can store data.

Reviewer 3: ```suggestion
Our fetcher will parse configuration to create Tasks and run them using [`FragmentsEngine`](https://github.com/Knotx/knotx-fragments/tree/2.0.0/handler/engine). It will output a `Fragment` that will contain data fetched from Google in its payload. Fragment's payload is just a `JsonObject` where actions can store data.
```

```java
private static final String FRAGMENT_TYPE = "graphql-data";

//...

private FragmentEventContextTaskAware setupTask(Vertx vertx, JsonObject config, RoutingContext routingContext, DataFetchingEnvironment env)  {
    JsonObject fragmentConfig = new JsonObject();
    fragmentConfig.put(FRAGMENT_TYPE, taskName);
    fragmentConfig.put("gql", new JsonObject(env.getArguments()));

    Fragment fragment = new Fragment(FRAGMENT_TYPE, fragmentConfig, "");

    RequestContext requestContext = routingContext.get(RequestContext.KEY);
    ClientRequest clientRequest = requestContext.getRequestEvent().getClientRequest();

    FragmentEvent event = new FragmentEvent(fragment);
    FragmentEventContext eventContext = new FragmentEventContext(event, clientRequest);

    FragmentsHandlerOptions options = new FragmentsHandlerOptions(config);
    ActionProvider proxyProvider = new ActionProvider(options.getAction(), supplyFactories(), vertx.getDelegate());
    TaskBuilder taskBuilder = new TaskBuilder(FRAGMENT_TYPE, options.getTasks(), proxyProvider);

    Task task = taskBuilder
        .build(fragment)

>         .orElseThrow(() -> new NullPointerException("No task built"));

Reviewer 4: Throwing NullPointerException by yourself? Seems to me that you should have your own ```IncorrectTaskInitializationException```.

Author: Yeah, it was a placeholder I forgot about. As by @tomaszmichalak suggestion, I've changed to `IllegalStateException`. 

I think this tutorial is big enough already, it's noted in the tutorial that a custom exception is a good idea, though

    return new FragmentEventContextTaskAware(task, eventContext);
}
```

Let's take a look at it piece by piece:

```java
JsonObject fragmentConfig = new JsonObject();
fragmentConfig.put(FRAGMENT_TYPE, taskName);
fragmentConfig.put("gql", new JsonObject(env.getArguments()));

Fragment fragment = new Fragment(FRAGMENT_TYPE, fragmentConfig, "");
```

We create a new fragment that will be processed by tasks. It needs to know which task should start processing it, therefore `taskName` is assigned to its type. We also put GraphQL environment arguments to the configuration under "gql" key. Environment arguments are just values passed to parameters of an operation. In our case "match" in `books(match: String)` and "id" in `book(id: String)`. By putting it into a fragment configuration, we allow tasks to access it. Note that the "gql" key is arbitrary and you can call it anything you want. 

> For the third argument in the fragment's constructor, we pass an empty string. It's our fragment's body which we're not interested in this tutorial.

Author: Leaving it as it is, like we discussed

```java
RequestContext requestContext = routingContext.get(RequestContext.KEY);
ClientRequest clientRequest = requestContext.getRequestEvent().getClientRequest();
```

We need to extract `ClientRequest` from `routingContext`. Remember the `${config.server.handlers.common.request}` part in the operation configuration in `operations.conf`? Without it there wouldn't be anything in `routingContext` under `RequestContext.KEY`.

```java
FragmentEvent event = new FragmentEvent(fragment);
FragmentEventContext eventContext = new FragmentEventContext(event, clientRequest);
```

> Our final task tree will need two things: `Task` and `FragmentEventContext`. Here we create the latter. It keeps the client request and our fragment in the form of `FragmentEvent`.

Reviewer 3: ```suggestion
Our final `Task` will need two things: Task's name and `FragmentEventContext`. Here we create the latter. It keeps the client request and our Fragment in the form of `FragmentEvent`.
```

```java
FragmentsHandlerOptions options = new FragmentsHandlerOptions(config);
ActionProvider proxyProvider = new ActionProvider(options.getAction(), supplyFactories(), vertx.getDelegate());
TaskBuilder taskBuilder = new TaskBuilder(FRAGMENT_TYPE, options.getTasks(), proxyProvider);
```

> In order to create a task we need to use a `TaskBuilder`. It needs to be supplied with a few things from the configuration. We use `FragmentsHandlerOptions` to parse the configuration. We also need a supplier of action factories. We can retrieve it using the standard java `ServiceLoader`. We'll take a look at `supplyFactories` method in a moment.

Reviewer 3: ```suggestion
In order to create a Task we need to use a `TaskBuilder`. It needs to be supplied with a few things from the configuration. We use `FragmentsHandlerOptions` to parse the configuration. We also need a supplier of action factories. We can retrieve it using the standard java `ServiceLoader`. We'll take a look at `supplyFactories` method in a moment.
```

Author: Leaving it, as discussed

    return () -> {
        ServiceLoader<ActionFactory> factories = ServiceLoader.load(ActionFactory.class);
        return factories.iterator();
    };
}
```

#### Retrieving payload
Now that we can make our fetcher execute Knot.x tasks from the configuration, we need to retrieve the outcome. As noted before, the data retrieved from the 3rd party (google books api in our case) is stored in the outcome fragment's payload. Therefore our final `get` method should look like this:

```java
@Override
public CompletableFuture<T> get(DataFetchingEnvironment environment) {
    FragmentEventContextTaskAware eventContextTaskAware = setupTask(vertx, config, routingContext, environment);
    CompletableFuture<T> future = new CompletableFuture<>();

    engine
        .execute(Collections.singletonList(eventContextTaskAware))
        .subscribe(events -> {
            JsonObject payload = events.get(0).getFragment().getPayload();
            JsonObject fetchedData = payload.getJsonObject("fetchedData");
            T model = getDataObjectFromJson(fetchedData, environment);
            future.complete(model);
        });

    return future;
}
```

> After the engine is done executing the task tree we:

Reviewer 3: ```suggestion
After the engine is done executing the task we:
```

```java
abstract T getDataObjectFromJson(JsonObject json, DataFetchingEnvironment environment) throws IllegalAccessException, InstantiationException;
```

#### Transforming json into model
Now we need to extend our `TaskDataFetcher` with something that can implement `getDataObjectFromJson` method. We'll create `TaskSingleDataFetcher` that'll transform json into a single model object, for example into `Book`.

```java
public class TaskSingleDataFetcher<T extends GraphQLDataObject> extends TaskDataFetcher<T> {
  private final Class<T> clazz;

  public TaskSingleDataFetcher(String task, Class<T> clazz, Vertx vertx, JsonObject config, RoutingContext routingContext) {
    super(vertx, config, routingContext, task);
    this.clazz = clazz;
  }

  @Override
  T getDataObjectFromJson(JsonObject json, DataFetchingEnvironment environment) throws IllegalAccessException, InstantiationException {
    T dataObject = clazz.newInstance();
    dataObject.fromJson(json, environment);
    return dataObject;
  }
}
```

> It's very simple. It takes a class of some object implementing `GraphQLDataObject` to be able to instantiate it. Remember that `Book` implements this interface and it has a single method: `fromJson` that takes a json and populates the object.

Reviewer 3: ```suggestion
It's very simple. It takes a class of some object implementing `GraphQLDataObject` to be able to instantiate it. Please note that it must provide also the default constructor. Remember that `Book` implements this interface and it has a single method: `fromJson` that takes a json and populates the object.
```

{
    "items": [
        {/* book 1 */},
        {/* book 2 */ },
        //...
    ]
}
```

Therefore, we would pass the following function as `toArray`: 

```java
json -> json.getJsonArray("items")
```

#### Using our fetchers
We can now assign our fetchers to operations in runtime wiring builder in `GraphQLHandlerFactory`:

```java
RuntimeWiring runtimeWiring = newRuntimeWiring()
        .type("QueryType", builder -> builder
            .dataFetcher("books", new TaskArrayDataFetcher<>("get-books", Book.class, vertx, config, routingContext, json -> json.getJsonArray("items")))
            .dataFetcher("book", new TaskSingleDataFetcher<>("get-book", Book.class, vertx, config, routingContext))
        )
        .build();
```

We just assign operation `books` to a `TaskArrayDataFetcher` that uses task `get-books` from configuration and returns an array of `Book` objects. Also, we assign operation `book` to a `TaskSingleDataFetcher` that uses task `get-book` and returns a `Book`.

Let's set it up in Knot.x configuration. It's best to extract our GraphQL configuration to a separate file. The GraphQL operation configuration in `operations.conf` should now look like this:

```HOCON
{

>     operationId = get-books-gql

Reviewer 3: remember about `-operation`.
